File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-variable-longest/longest-substring-without-repeating/code/cpp/longest_substring_without_repeating_sol2.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> window;
        int left = 0, maxLen = 0;
        for (int right = 0; right < s.size(); right++) {
            while (window.count(s[right]))  // Shrink until no repeat
                window.erase(s[left++]);
            window.insert(s[right]);
            maxLen = max(maxLen, right - left + 1);
        }
        return maxLen;
    }
};

File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-variable-longest/longest-substring-without-repeating/code/python/longest_substring_without_repeating_sol2.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        window, left, max_len = set(), 0, 0
        for right in range(len(s)):
            while s[right] in window: window.remove(s[left]); left += 1
            window.add(s[right]); max_len = max(max_len, right-left+1)
        return max_len